File: apps/gain.cpp
// Project Name: OpenMEEG (http://openmeeg.github.io)
// © INRIA and ENPC under the French open source license CeCILL-B.
// See full copyright notice in the file LICENSE.txt
// If you make a copy of this file, you must either:
// - provide also LICENSE.txt and modify this header to refer to it.
// - replace this header by the LICENSE.txt content.

#include <om_utils.h>
#include <commandline.h>
#include <gain.h>

using namespace OpenMEEG;

void help(const char* cmd_name);

inline void
error(const char* command,const bool unknown_option=false) {
    std::cerr << "Error: " << ((unknown_option) ? "Unknown option." : "Not enough arguments.") << std::endl
              << "Please try \"" << command << " -h\" or \"" << command << " --help \" \n" << std::endl;
    exit(1);
}

int
main(int argc,char** argv) {

    const CommandLine cmd(argc,argv);

    if (cmd.help_mode()) {
        help(argv[0]);
        return 0;
    }

    if (argc<5) {
        std::cerr << "Not enough arguments." << std::endl;
        help(argv[0]);
        return 1;
    }

    print_version(argv[0]);
    cmd.print();

    constexpr char geomfileopt[]       = "geometry file";
    constexpr char condfileopt[]       = "conductivity file";
    constexpr char outputfileopt[]     = "output file";
    constexpr char dipolefileopt[]     = "dipoles file";
    constexpr char HMfileopt[]         = "head matrix";
    constexpr char invHMfileopt[]      = "inverse head matrix";
    constexpr char sourcematfileopt[]  = "source matrix";
    constexpr char h2eegfileopt[]      = "head to EEG matrix";
    constexpr char h2megfileopt[]      = "head to MEG matrix";
    constexpr char h2ipfileopt[]       = "head to internal potential matrix";
    constexpr char source2megfileopt[] = "sources to MEG matrix";
    constexpr char source2ipfileopt[]  = "sources to internal potential matrix";

    // Start Chrono

    const auto start_time = std::chrono::system_clock::now();

    unsigned num_options = 0;

    const auto& EEGparms = {
        invHMfileopt, sourcematfileopt, h2eegfileopt, outputfileopt
    };

    if (char** opt_parms = cmd.option("-EEG",EEGparms)) {

        assert_non_conflicting_options(argv[0],++num_options);

        // EEG DATA

        //  Split the 2 matrix multiplications in order to spare memory.
        //  This is why we do not use GainEEG...

        const SymMatrix    HeadMatInv(opt_parms[1]);
        const SparseMatrix Head2EEGMat(opt_parms[3]);
        const Matrix& tmp = Head2EEGMat*HeadMatInv;
        const Matrix SourceMat(opt_parms[2]);
        const Matrix& EEGGainMat = tmp*SourceMat;
        EEGGainMat.save(opt_parms[4]);
    }

    const auto& EEGAdjointparms = {
        geomfileopt, condfileopt, dipolefileopt, HMfileopt, h2eegfileopt, outputfileopt
    };

    if (char** opt_parms = cmd.option("-EEGadjoint",EEGAdjointparms)) {

        assert_non_conflicting_options(argv[0],++num_options);

        // Compute gain matrix with the adjoint method for use with EEG DATA

        Geometry geo(opt_parms[1],opt_parms[2]);
        const Matrix dipoles(opt_parms[3]);
        const SymMatrix HeadMat(opt_parms[4]);
        const SparseMatrix Head2EEGMat(opt_parms[5]);

        const GainEEGadjoint EEGGainMat(geo, dipoles, HeadMat, Head2EEGMat);
        EEGGainMat.save(opt_parms[6]);
    }

    const auto& MEGparms = {
        invHMfileopt, sourcematfileopt, h2megfileopt, source2megfileopt, outputfileopt
    };

    if (char** opt_parms = cmd.option("-MEG",MEGparms)) {

        assert_non_conflicting_options(argv[0],++num_options);

        // MEG DATA

        //  We split the 3 matrix multiplications in order to spare memory.
        //  This is also why we do not use GainMEG...

        const SymMatrix HeadMatInv(opt_parms[1]);
        const Matrix Head2MEGMat(opt_parms[3]);
        const Matrix& tmp1 = Head2MEGMat*HeadMatInv;
        const Matrix SourceMat(opt_parms[2]);
        const Matrix& tmp2 = tmp1*SourceMat;
        const Matrix Source2MEGMat(opt_parms[4]);
        const Matrix MEGGainMat = Source2MEGMat+tmp2;
        MEGGainMat.save(opt_parms[5]);
    }

    const auto& MEGAdjointparms = {
        geomfileopt, condfileopt, dipolefileopt, HMfileopt, h2megfileopt, source2megfileopt, outputfileopt
    };

    if (char** opt_parms = cmd.option("-MEGadjoint",MEGAdjointparms)) {

        assert_non_conflicting_options(argv[0],++num_options);

        // Compute the gain matrix with the adjoint method for use with MEG DATA

        Geometry geo(opt_parms[1],opt_parms[2]);
        const Matrix dipoles(opt_parms[3]);
        const SymMatrix HeadMat(opt_parms[4]);
        const Matrix Head2MEGMat(opt_parms[5]);
        const Matrix Source2MEGMat(opt_parms[6]);

        const GainMEGadjoint MEGGainMat(geo, dipoles, HeadMat, Head2MEGMat, Source2MEGMat);
        MEGGainMat.save(opt_parms[7]);
    }

    const auto& EEGMEGAdjointparms = {
        geomfileopt, condfileopt, dipolefileopt, HMfileopt, h2eegfileopt, h2megfileopt, source2megfileopt,
        "EEG output file", "MEG output file"
    };

    if (char** opt_parms = cmd.option("-EEGMEGadjoint",EEGMEGAdjointparms)) {

        assert_non_conflicting_options(argv[0],++num_options);

        // Compute the gain matrices with the adjoint method for EEG and MEG DATA

        Geometry geo(opt_parms[1],opt_parms[2]);
        const Matrix dipoles(opt_parms[3]);
        const SymMatrix HeadMat(opt_parms[4]);
        const SparseMatrix Head2EEGMat(opt_parms[5]);
        const Matrix Head2MEGMat(opt_parms[6]);
        const Matrix Source2MEGMat(opt_parms[7]);

        const GainEEGMEGadjoint EEGMEGGainMat(geo,dipoles,HeadMat,Head2EEGMat,Head2MEGMat,Source2MEGMat);
        EEGMEGGainMat.saveEEG(opt_parms[8]);
        EEGMEGGainMat.saveMEG(opt_parms[9]);
    }

    const auto& IPparms = {
        invHMfileopt, sourcematfileopt, h2ipfileopt, source2ipfileopt, outputfileopt
    };

    if (char** opt_parms = cmd.option({ "-InternalPotential", "-IP", "-ip" },IPparms)) {

        assert_non_conflicting_options(argv[0],++num_options);

        const SymMatrix HeadMatInv(opt_parms[1]);
        const Matrix Head2IPMat(opt_parms[3]);

        const Matrix& tmp1 = Head2IPMat*HeadMatInv;
        const Matrix SourceMat(opt_parms[2]);
        const Matrix& tmp2 = tmp1*SourceMat;
        const Matrix Source2IPMat(opt_parms[4]);

        const Matrix& InternalPotGainMat = Source2IPMat+tmp2;
        InternalPotGainMat.save(opt_parms[5]);
    }

    const auto& EITIPparms = {
        invHMfileopt, sourcematfileopt, h2ipfileopt, outputfileopt
    };

    if (char** opt_parms = cmd.option({ "-EITInternalPotential", "-EITIP", "-eitip" },EITIPparms)) {

        assert_non_conflicting_options(argv[0],++num_options);

        const SymMatrix HeadMatInv(opt_parms[1]);
        const Matrix SourceMat(opt_parms[2]);
        const Matrix Head2IPMat(opt_parms[3]);

        const Matrix& InternalPotGainMat = (Head2IPMat*HeadMatInv)*SourceMat;

        InternalPotGainMat.save(opt_parms[4]);
    }

    if (num_options==0) {
        std::cerr << "Unknown argument: " << argv[1] << std::endl;
        exit(1);
    }

    // Stop Chrono

    const auto end_time = std::chrono::system_clock::now();
    dispEllapsed(end_time-start_time);

    return 0;
}

void
help(const char* cmd_name) {

    std::cout << cmd_name << " [-option] [filepaths...]" << std::endl << std::endl;

    std::cout << "-option :" << std::endl
              << "   -EEG :   Compute the gain for EEG " << std::endl
              << "            Filepaths are in order :" << std::endl
              << "            HeadMatInv, SourceMat, Head2EEGMat, EEGGainMatrix" << std::endl
              << "            bin Matrix" << std::endl << std::endl;

    std::cout << "   -MEG :   Compute the gain for MEG " << std::endl
              << "            Filepaths are in order :" << std::endl
              << "            HeadMatInv, SourceMat, Head2MEGMat, Source2MEGMat, MEGGainMatrix" << std::endl
              << "            Matrix (.bin or .txt)" << std::endl << std::endl;

    std::cout << "   -InternalPotential or -IP :   Compute the gain for internal potentials, measured within the volume " << std::endl
              << "            Filepaths are in order :" << std::endl
              << "            HeadMatInv, SourceMat, Head2IPMat, Source2IPMat" << std::endl
              << "            InternalPotential gain Matrix (.txt)" << std::endl << std::endl;

    std::cout << "   -EITInternalPotential or -EITIP :   Compute the gain for internal potentials using boundary normal current as input" << std::endl
              << "            Filepaths are in order :" << std::endl
              << "            HeadMatInv, SourceMat, Head2IPMat" << std::endl
              << "            EITInternalPotential gain Matrix" << std::endl << std::endl;

    std::cout << "   -EEGadjoint :   Compute the gain for EEG " << std::endl
              << "            Filepaths are in order :" << std::endl
              << "            geometry file (.geom)" << std::endl
              << "            conductivity file (.cond)" << std::endl
              << "            dipoles positions and orientations" << std::endl
              << "            HeadMat, Head2EEGMat, EEGGainMatrix" << std::endl
              << "            bin Matrix" << std::endl << std::endl;

    std::cout << "   -MEGadjoint :   Compute the gain for MEG " << std::endl
              << "            Filepaths are in order :" << std::endl
              << "            geometry file (.geom)" << std::endl
              << "            conductivity file (.cond)" << std::endl
              << "            dipoles positions and orientations" << std::endl
              << "            HeadMat, Head2MEGMat, Source2MEGMat, MEGGainMatrix" << std::endl
              << "            bin Matrix" << std::endl << std::endl;

    std::cout << "   -EEGMEGadjoint :   Compute the gain for EEG and MEG at the same time" << std::endl
              << "            Filepaths are in order :" << std::endl
              << "            geometry file (.geom)" << std::endl
              << "            conductivity file (.cond)" << std::endl
              << "            dipoles positions and orientations" << std::endl
              << "            HeadMat, Head2EEGMat, Head2MEGMat, Source2MEGMat, EEGGainMatrix, MEGGainMatrix" << std::endl
              << "            bin Matrix" << std::endl << std::endl;
}

File: wrapping/python/apps/gain.py
import os
import sys
import time
import argparse
import pathlib
import openmeeg as om

def parse_arguments(argv):
    parser = argparse.ArgumentParser(description='Compute various leadfields (gain matrices).')
    subparsers = parser.add_subparsers(dest='subparser_name', help='sub-command help')

    # create the parser for the EEG case
    EEGparser = subparsers.add_parser('EEG', help='Compute the EEG gain matrix')
    EEGparser.add_argument('HeadMatInv', type=open, nargs=1, help='file name containing the inverse of the head matrix')
    EEGparser.add_argument('SourceMat', type=open, nargs=1, help='file name containing the source matrix')
    EEGparser.add_argument('Head2EEGMat', type=open, nargs=1, help='file name containing the head to EEG conversion matrix')
    EEGparser.add_argument('dest_file', metavar='EEGGainMatrix', type=str, nargs=1, help='output: EEG gain matrix file name')

    # create the parser for the MEG case
    MEGparser = subparsers.add_parser('MEG', help='Compute the MEG gain matrix')
    MEGparser.add_argument('HeadMatInv', type=open, nargs=1, help='file name containing the inverse of the head matrix')
    MEGparser.add_argument('SourceMat', type=open, nargs=1, help='file name containing the source matrix')
    MEGparser.add_argument('Head2MEGMat', type=open, nargs=1, help='file name containing the head to MEG conversion matrix')
    MEGparser.add_argument('Source2MEGMat', type=open, nargs=1, help='file name containing the source to MEG conversion matrix')
    MEGparser.add_argument('dest_file', metavar='MEGGainMatrix', type=str, nargs=1, help='output: MEG gain matrix file name')

    # create the parser for the internal potential case
    IPparser = subparsers.add_parser('InternalPotential', aliases=['IP'], help='Compute the internal potential gain matrix')
    IPparser.add_argument('HeadMatInv', type=open, nargs=1, help='file name containing the inverse of the head matrix')
    IPparser.add_argument('SourceMat', type=open, nargs=1, help='file name containing the source matrix')
    IPparser.add_argument('Head2IPMat', type=open, nargs=1, help='file name containing the head to internal potential conversion matrix')
    IPparser.add_argument('Source2IPMat', type=open, nargs=1, help='file name containing the source to internal potential conversion matrix')
    IPparser.add_argument('dest_file', metavar='IPGainMatrix', type=str, nargs=1, help='output: EEG gain matrix file name')

    # create the parser for the EIT internal potential case
    EITIPparser = subparsers.add_parser('EITInternalPotential', aliases=['EITIP'], help='Compute the gain matrix for internal potentials using boundary normal current as input')
    EITIPparser.add_argument('HeadMatInv', type=open, nargs=1, help='file name containing the inverse of the head matrix')
    EITIPparser.add_argument('SourceMat', type=open, nargs=1, help='file name containing the source matrix')
    EITIPparser.add_argument('Head2IPMat', type=open, nargs=1, help='file name containing the head to internal potential conversion matrix')
    EITIPparser.add_argument('dest_file', metavar='EEGGainMatrix', type=str, nargs=1, help='output: EEG gain matrix file name')

    # create the parser for the EEG adjoint case
    EEGAdjointparser = subparsers.add_parser('EEGadjoint', help='Compute the EEG gain matrix')
    EEGAdjointparser.add_argument('geometry', type=open, nargs=1, help='head geometry file name (.geom)')
    EEGAdjointparser.add_argument('conductivities', type=open, nargs=1, help='head conductivities file name (.cond)')
    EEGAdjointparser.add_argument('sources', type=open, nargs=1, help='file name containing the description of the dipoles (positions, orientations)')
    EEGAdjointparser.add_argument('HeadMat', type=open, nargs=1, help='file name containing the head matrix')
    EEGAdjointparser.add_argument('Head2EEGMat', type=open, nargs=1, help='output: file name containing the head to EEG conversion matrix')
    EEGAdjointparser.add_argument('dest_file', metavar='EEGGainMatrix', type=str, nargs=1, help='output: EEG gain matrix file name')

    # create the parser for the MEG adjoint case
    MEGAdjointparser = subparsers.add_parser('MEGadjoint', help='Compute the MEG gain matrix')
    MEGAdjointparser.add_argument('geometry', type=open, nargs=1, help='head geometry file name (.geom)')
    MEGAdjointparser.add_argument('conductivities', type=open, nargs=1, help='head conductivities file name (.cond)')
    MEGAdjointparser.add_argument('sources', type=open, nargs=1, help='file name containing the description of the dipoles (positions, orientations)')
    MEGAdjointparser.add_argument('HeadMat', type=open, nargs=1, help='file name containing the head matrix')
    MEGAdjointparser.add_argument('Head2MEGMat', type=open, nargs=1, help='file name containing the head to MEG conversion matrix')
    MEGAdjointparser.add_argument('Source2MEGMat', type=open, nargs=1, help='file name containing the source to MEG conversion matrix')
    MEGAdjointparser.add_argument('dest_file', metavar='MEGGainMatrix', type=str, nargs=1, help='output: MEG gain matrix file name')

    # create the parser for the joint EEG MEG adjoint case
    EEGMEGAdjointparser = subparsers.add_parser('EEGMEGadjoint', help='Compute both the EEG and MEG gain matrices')
    EEGMEGAdjointparser.add_argument('geometry', type=open, nargs=1, help='head geometry file name (.geom)')
    EEGMEGAdjointparser.add_argument('conductivities', type=open, nargs=1, help='head conductivities file name (.cond)')
    EEGMEGAdjointparser.add_argument('sources', type=open, nargs=1, help='file name containing the description of the dipoles (positions, orientations)')
    EEGMEGAdjointparser.add_argument('HeadMat', type=open, nargs=1, help='file name containing the head matrix')
    EEGMEGAdjointparser.add_argument('Head2EEGMat', type=open, nargs=1, help='file name containing the head to EEG conversion matrix')
    EEGMEGAdjointparser.add_argument('Head2MEGMat', type=open, nargs=1, help='file name containing the head to MEG conversion matrix')
    EEGMEGAdjointparser.add_argument('Source2MEGMat', type=open, nargs=1, help='file name containing the source to MEG conversion matrix')
    EEGMEGAdjointparser.add_argument('EEG_dest_file', metavar='EEGGainMatrix', type=str, nargs=1, help='output: EEG gain matrix file name')
    EEGMEGAdjointparser.add_argument('MEG_dest_file', metavar='MEGGainMatrix', type=str, nargs=1, help='output: MEG gain matrix file name')

    args = parser.parse_args(argv[1:])
    args.processing_function = args.subparser_name

    return args

def EEG(args):
    HeadMatInv  = om.SymMatrix(args.HeadMatInv)
    Head2EEGMat = om.SparseMatrix(args.Head2EEGMat)
    tmp         = Head2EEGMat*HeadMatInv
    SourceMat   = om.Matrix(args.SourceMat)
    EEGGainMat  = tmp*SourceMat
    EEGGainMat.save(args.dest_file)

def EEGadjoint(args):
    geometry    = om.Geometry(args.geometry,args.conductivities)
    dipoles     = om.Matrix(args.sources)
    HeadMat     = om.SymMatrix(args.HeadMat)
    Head2EEGMat = om.SparseMatrix(args.Head2EEGMat)
    EEGGainMat  = om.GainEEGadjoint(geometry, dipoles, HeadMat, Head2EEGMat)
    EEGGainMat.save(args.dest_file)

def MEG(args):
    HeadMatInv    = om.SymMatrix(args.HeadMatInv)
    Head2MEGMat   = om.Matrix(args.Head2MEGMat)
    tmp1          = Head2MEGMat*HeadMatInv
    SourceMat     = om.Matrix(args.SourceMat)
    tmp2          = tmp1*SourceMat
    Source2MEGMat = om.Matrix(args.Source2MEGMat)
    MEGGainMat    = Source2MEGMat+tmp2
    MEGGainMat.save(args.dest_file)

def MEGadjoint(args):
    geometry      = om.Geometry(args.geometry,args.conductivities)
    dipoles       = om.Matrix(args.sources)
    HeadMat       = om.SymMatrix(args.HeadMat)
    Head2MEGMat   = om.Matrix(args.Head2MEGMat)
    Source2MEGMat = om.Matrix(args.Source2MEGMat)
    MEGGainMat    = om.GainMEGadjoint(geometry, dipoles, HeadMat, Head2MEGMat, Source2MEGMat)
    MEGGainMat.save(args.dest_file)

def EEGMEGadjoint(args):
    geometry      = om.Geometry(args.geometry,args.conductivities)
    dipoles       = om.Matrix(args.sources)
    HeadMat       = om.SymMatrix(args.HeadMat)
    Head2EEGMat   = om.SparseMatrix(args.Head2EEGMat)
    Head2MEGMat   = om.Matrix(args.Head2MEGMat)
    Source2MEGMat = om.Matrix(args.Source2MEGMat)
    EEGMEGGainMat = om.EEGMEGGainMat(geometry, dipoles, HeadMat, Head2EEGMat, Head2MEGMat, Source2MEGMat)
    EEGMEGGainMat.saveEEG(args.EEG_dest_file)
    EEGMEGGainMat.saveMEG(args.MEG_dest_file)

def InternalPotential(args):
    HeadMatInv         = om.SymMatrix(args.HeadMatInv)
    Head2IPMat         = om.SparseMatrix(args.Head2IPMat)
    tmp1               = Head2IPMat*HeadMatInv
    SourceMat          = om.Matrix(args.SourceMat)
    tmp2               = tmp1*SourceMat
    Source2IPMat       = om.Matrix(args.Source2IPMat)
    InternalPotGainMat = Source2IPMat+tmp2
    InternalPotGainMat.save(args.dest_file)

def EITInternalPotential(args):
    HeadMatInv         = om.SymMatrix(args.HeadMatInv)
    SourceMat          = om.Matrix(args.SourceMat)
    Head2IPMat         = om.SparseMatrix(args.Head2IPMat)
    InternalPotGainMat = (Head2IPMat*HeadMatInv)*SourceMat
    InternalPotGainMat.save(args.dest_file)
